### PR TITLE
Feature/graphql wishlist remove item

### PR DESCRIPTION
**Pull request by LucasCalazans**
### Description

Adding a new mutation to the WishlistGraphQl to remove multiple items

### Fixed Issues
1. #849: [Wish List] Remove products from the Wish List

### Manual testing scenarios
1. Use the new mutation to remove a previously added item from your wishlist passing the products ids (It can be one or many)
```graphql
mutation wishlistRemoveItem($wishlist_id: ID!, $wishlist_items_ids: [ID!]!) {
  removeProductsFromWishlist(input: {
    wishlist_id: $wishlist_id
    wishlist_items_ids: $wishlist_items_ids
  }) {
    wishlist {
      id
      items {
        id
        product {
          id
          name
        }	
      }
      items_count
      sharing_code
      updated_at
    }
  }
}
```

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/WishlistGraphQl/etc/schema.graphqls [hunk 1]
```diff
@@ -5,6 +5,10 @@ type Query {
     wishlist: WishlistOutput @resolver(class: "\\Magento\\WishlistGraphQl\\Model\\Resolver\\WishlistResolver") @doc(description: "The wishlist query returns the contents of a customer's wish list") @cache(cacheable: false)
 }
 
+type Mutation {
+    removeProductsFromWishlist(input: [Int]!): WishlistOutput @resolver(class: "\\Magento\\WishlistGraphQl\\Model\\Resolver\\WishlistRemoveItem") @doc(description: "Delete a list of wishlist items")
```
Comment (lenaorobei) [inline app/code/Magento/WishlistGraphQl/etc/schema.graphqls]: As per design document, it should be: https://github.com/magento/architecture/pull/326
```suggestion
    removeProductsFromWishlist(input: RemoveProductsFromWishlistInput): RemoveProductsFromWishlistOutput @resolver(class: "\\Magento\\WishlistGraphQl\\Model\\Resolver\\WishlistRemoveItem") @doc(description: "Delete a list of wishlist items")
```

**Review by lenaorobei — CHANGES_REQUESTED** [review]
Thank you for the contribution!. Please use schema from the approved design document https://github.com/magento/architecture/pull/326 and cover this functionality with api-functional tests.

**Comment by LucasCalazans** [comment]
It's just missing the API functional tests. I'll try to do that this weekend.

File: app/code/Magento/WishlistGraphQl/etc/schema.graphqls [hunk 2]
```diff
@@ -5,16 +5,12 @@ type Query {
     wishlist: WishlistOutput @resolver(class: "\\Magento\\WishlistGraphQl\\Model\\Resolver\\WishlistResolver") @deprecated(reason: "Moved under `Customer` `wishlist`") @doc(description: "The wishlist query returns the contents of a customer's wish list") @cache(cacheable: false)
 }
 
-type Customer {
-    wishlist: Wishlist! @resolver(class:"\\Magento\\WishlistGraphQl\\Model\\Resolver\\CustomerWishlistResolver") @doc(description: "The wishlist query returns the contents of a customer's wish lists")  @cache(cacheable: false)
```
Comment (lenaorobei) [inline app/code/Magento/WishlistGraphQl/etc/schema.graphqls]: `Customer` `wishlist` should not be removed.

**Comment by lenaorobei** [comment]
@magento run all tests

File: app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        ];
+    }
+
+    /**
+     * Loads Wishlist by the wishlist id
+     *
+     * @param int $wishlistId
+     * @return Wishlist
+     */
+    private function getWishlistById($wishlistId): Wishlist
```
Comment (lenaorobei) [inline app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php]: This code is also used in `\Magento\WishlistGraphQl\Model\Resolver\CustomerWishlistResolver::resolve`. We need to move to the separate class in the `execute` method and reuse this code. Example: `\Magento\CustomerGraphQl\Model\Customer\SaveCustomer::execute`

File: app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+     */
+    private function deleteWishlistItems(Wishlist $wishlist, Array $ids)
+    {
+        $wishlistItems = $this->getWishListItems($wishlist, $ids);
+
+        foreach ($wishlistItems as $whislistItem) {
+            try {
+                $this->wishlistItemResource->delete($whislistItem);
+            } catch (\Exception $e) {
+                throw new LocalizedException(__('Failed to delete the wishlist item'));
```
Comment (lenaorobei) [inline app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php]: `GraphQlInputException` should be thrown.

File: app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+     * @param array $ids
+     * @return array
+     */
+    private function getWishListItems(Wishlist $wishlist, Array $ids): array
+    {
+        /** @var WishlistItemCollection $wishlistItemCollection */
+        $wishlistItemCollection = $this->wishlistItemCollectionFactory->create();
+        $wishlistItemCollection
+            ->addWishlistFilter($wishlist)
+            ->addStoreFilter(array_map(function (StoreInterface $store) {
```
Comment (lenaorobei) [inline app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php]: We should get `store` from $context.
`$storeId = (int)$context->getExtensionAttributes()->getStore()->getId();`

**Comment by TomashKhamlai** [comment]
Steps to reproduce:
1. Install Magento.
2. Don't register customer
3. Don't create any products
4. Don't add any items to wishlist
5. Send request:
#### Request:
```graphql
mutation wishlistRemoveItem(
  $wishlist_id: ID!
  $wishlist_items_ids: [ID!]!
) {
  removeProductsFromWishlist(
    input: {
      wishlist_id: $wishlist_id
      wishlist_items_ids: $wishlist_items_ids
    }
  ) {
    wishlist {
      id
      items {
        id
        product {
          id
          name
        }
      }
      items_count
      sharing_code
      updated_at
    }
  }
}
```
#### Variables
```json
{
  "wishlist_id": "1",
  "wishlist_items_ids": ["2", "3"]
}
```
### Expected result:
Some of the errors:
 - Customer is not authorized to perform operations on wishlist with id=1
 - No such entity exception
### Actual result:
```graphql
{
  "data": {
    "removeProductsFromWishlist": {
      "wishlist": {
        "id": null,
        "items": [],
        "items_count": 0,
        "sharing_code": null,
        "updated_at": null
      }
    }
  }
}
```

File: app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php (hunk 3 above)
Comment (LucasCalazans) [reply app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php]: Hey @lenaorobei it is just missing this now.

I didn't understand why we should create a `Customer\SaveCustomer` to add the `getWishlistById` method. They are not related, actually.

Can you explain I little bit more about it? I was thinking to add it to a new class in the `WishlistGraphQl\Model` folder, but naming it as something with Wishlist in the name.

**Comment by LucasCalazans** [comment]
@TomashKhamlai Hey, thank you for your review. I've made some fixies, can you check again, please?

File: app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php (hunk 3 above)
Comment (lenaorobei) [reply app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php]: I meant `Magento\CustomerGraphQl\Model\Customer\SaveCustomer::execute` as an example. In our case it should be `GetWishlistById` class and `execute` method.

**Comment by TomashKhamlai** [comment]
@LucasCalazans, please pull the latest changes on 2.3-develop, merge them into your branch and push the changes. I am blocked by:
```
main.CRITICAL: Class Magento\Catalog\Model\Product\Attribute\Source\LayoutUpdate does not exist {"report_id":"da73692f7a69afb4559ae3ee4b072292ba89f137f0cf4b249c23fdb39a0902a3","exception":"[object] (ReflectionException(code: -1): Class Magento\\Catalog\\Model\\Product\\Attribute\\Source\\LayoutUpdate does not exist at /var/www/html/graphqlce/lib/internal/Magento/Framework/Code/Reader/ClassReader.php:26)"} []
main.CRITICAL: Unable to serialize value. Error: Malformed UTF-8 characters, possibly incorrectly encoded {"exception":"[object] (InvalidArgumentException(code: 0): Unable to serialize value. Error: Malformed UTF-8 characters, possibly incorrectly encoded at /var/www/html/graphqlce/lib/internal/Magento/Framework/Serialize/Serializer/Json.php:26)"} []
```
Merging the latest mainline is a known and verified troubleshooting step for the problem with missing  `LayoutUpdate.php` file

**Comment by TomashKhamlai** [comment]
There is another problem:

### Steps to reproduce:
1. Create customer
2. Don't create products so there is nothing to add to wishlist
3. Generate Customer Token
4. Use query `customer` to get wishlist's `id`
5.  Send a request with valid data(use any integer numbers that are greater than 0 for `wishlist_items_ids`):
#### Request:
```graphql
mutation wishlistRemoveItem(
  $wishlist_id: ID!
  $wishlist_items_ids: [ID!]!
) {
  removeProductsFromWishlist(
    input: {
      wishlist_id: $wishlist_id
      wishlist_items_ids: $wishlist_items_ids
    }
  ) {
    wishlist {
      id
      items {
        id
        product {
          id
          name
        }
      }
      items_count
      sharing_code
      updated_at
    }
  }
}
```
#### Variables
```json
{
  "wishlist_id": "< valid_id >",
  "wishlist_items_ids": ["2", "3"]
}
```
#### Variables
```json
{
  "Authorization": "Bearer < valid_token >"
}
```
6. Open another tab and generate the customer's token using the wrong password 10 times in a row.
7. Repeat step 5.

### Expected result:
The locked customer is not able to perform this request and is notified with some message about it.

### Actual result:
No difference between the locked and active customer
```
{
  "data": {
    "removeProductsFromWishlist": {
      "wishlist": {
        "id": "6",
        "items": [],
        "items_count": 0,
        "sharing_code": "yEbOzoRaadxxIn5YlofKYSUkUxufVGKD",
        "updated_at": "2019-11-15 17:33:28"
      }
    }
  }
}
```

**Comment by TomashKhamlai** [comment]
@chalov-anton, please continue testing this PR.

**Comment by chalov-anton** [comment]
> Steps to reproduce:
> 
> 1. Install Magento.
> 2. Don't register customer
> 3. Don't create any products
> 4. Don't add any items to wishlist
> 5. Send request:
> 
> #### Request:
> ```graphql
> mutation wishlistRemoveItem(
>   $wishlist_id: ID!
>   $wishlist_items_ids: [ID!]!
> ) {
>   removeProductsFromWishlist(
>     input: {
>       wishlist_id: $wishlist_id
>       wishlist_items_ids: $wishlist_items_ids
>     }
>   ) {
>     wishlist {
>       id
>       items {
>         id
>         product {
>           id
>           name
>         }
>       }
>       items_count
>       sharing_code
>       updated_at
>     }
>   }
> }
> ```
> 
> #### Variables
> ```json
> {
>   "wishlist_id": "1",
>   "wishlist_items_ids": ["2", "3"]
> }
> ```
> 
> ### Expected result:
> Some of the errors:
> 
> * Customer is not authorized to perform operations on wishlist with id=1
> * No such entity exception
> 
> ### Actual result:
> ```graphql
> {
>   "data": {
>     "removeProductsFromWishlist": {
>       "wishlist": {
>         "id": null,
>         "items": [],
>         "items_count": 0,
>         "sharing_code": null,
>         "updated_at": null
>       }
>     }
>   }
> }
> ```

If the Customer is **not** Authorized, for example the Authorization token is wrong(or not provided), an error message appears after performing mutation :thumbsup:
```graphql
{
  "errors": [
    {
      "message": "The current user cannot perform operations on wishlist",
      "extensions": {
        "category": "graphql-authorization"
      },
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "removeProductsFromWishlist"
      ]
    }
  ],
  "data": {
    "removeProductsFromWishlist": null
  }
}
```
But still, if the customer is authorized and trying to perform operations on empty wishlists - a "success" message returns 
```graphql
{
  "data": {
    "removeProductsFromWishlist": {
      "wishlist": {
        "id": "1",
        "items": [],
        "items_count": 0,
        "sharing_code": "4zNIfYqgRRxpzTnDWEP2ZoPKmwbSuy15",
        "updated_at": "2019-11-18 08:10:16"
      }
    }
  }
} 
```
Expected error message like `There is no %s item in wishlist`.

File: app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php (hunk 3 above)
Comment (LucasCalazans) [reply app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php]: Hmm, I see. I'll do that.

**Comment by LucasCalazans** [comment]
@chalov-anton I thought that it should return everything and ignore the ids.

Okay, when the user passes 2 ids. When the first exists and another one doesn't. What should I do in this case? Show an error or return the success object?

**Comment by lenaorobei** [comment]
@LucasCalazans the general rule for GraphQL functionality test coverage is to use api-functional tests. You can find some examples here `/dev/tests/api-functional/testsuite/Magento/GraphQl`. Unit tests do not cover actual use case, so please consider removing them and write api-functional tests. Thanks.

File: app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    public function resolve(
+        Field $field,
+        $context,
+        ResolveInfo $info,
+        array $value = null,
+        array $args = null
+    ) {
+        $customerId = $context->getUserId();
+
+        if (!$customerId && $customerId === 0) {
```
Comment (lenaorobei) [inline app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php]:
```
if (false === $context->getExtensionAttributes()->getIsCustomer()) {
    throw new GraphQlAuthorizationException(__('The current customer isn\'t authorized.'));
}
```

File: app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+            throw new GraphQlAuthorizationException(__('The current user cannot perform operations on wishlist'));
+        }
+
+        $wishlist = $this->getWishlistByAttribute->execute('wishlist_id', $args['input']['wishlist_id']);
+
+        if ((int) $wishlist->getCustomerId() !== $customerId) {
+            throw new GraphQlAuthorizationException(__('The current user doesn\'t have a wishlist with the provided id'));
```
Comment (lenaorobei) [inline app/code/Magento/WishlistGraphQl/Model/Resolver/WishlistRemoveItemResolver.php]: Let's do not expose additional info. Message should be the same `throw new GraphQlAuthorizationException(__('The current customer isn\'t authorized.'));`

**Review by lenaorobei — CHANGES_REQUESTED** [review]
Please address review comments and add api-functional tests.

**Comment by LucasCalazans** [comment]
@lenaorobei I'll need more time to create the api-functional tests because I have never used it.

**Comment by lenaorobei** [comment]
@LucasCalazans please let me know if you need any help with this.
Docs: https://devdocs.magento.com/guides/v2.3/graphql/functional-testing.html
Example: `Magento\GraphQl\Quote\Guest\RemoveItemFromCartTest`.

**Comment by LucasCalazans** [comment]
Okay, I'll try to do that today.

@lenaorobei there are some errors in the Semantic Version Checker and Unit Tests. Do you know why? Nothing in this PR was changed related to this.

**Comment by lenaorobei** [comment]
@LucasCalazans, looks like an issue on our side. Let's ignore for now, I'll look into it.

**Comment by chalov-anton** [comment]
There is another issue with this PR
If perform the mutation with **correct** Authorization Header but with **wrong** Variables, for example:
```graphql
{
  "wishlist_id": "non_existing",
  "wishlist_items_ids": ["non_existing"]
}
```
The response message is 
```graphql
{
  "errors": [
    {
      "message": "The current customer isn't authorized.",
      "extensions": {
        "category": "graphql-authorization"
      },
...
```
In this case, we would expect an error message about wrong variables or _Non-existing wishlist_id_

**Comment by lenaorobei** [comment]
Hi @LucasCalazans! Unfortunately we need to close this PR in order to not to increase tech debt with new storefront approach.
Hope you'll join us in upcoming project.

**Pull request closed without merging**